hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/container/common/helpers/ExcludeList.java

```diff
@@ -34,31 +37,54 @@
  */
 public class ExcludeList {
 
-  private final Set<DatanodeDetails> datanodes;
+  private final Map<DatanodeDetails, Long> datanodes;
   private final Set<ContainerID> containerIds;
   private final Set<PipelineID> pipelineIds;
+  private long expiryTime = 0;
+  private java.time.Clock clock;
 
 
   public ExcludeList() {
-    datanodes = new HashSet<>();
+    datanodes = new HashMap<>();
     containerIds = new HashSet<>();
     pipelineIds = new HashSet<>();
   }
 
+  public ExcludeList(long autoExpiryTime, java.time.Clock clock) {
+    this();
+    this.expiryTime = autoExpiryTime;
+    this.clock = clock;
+  }
+
   public Set<ContainerID> getContainerIds() {
     return containerIds;
   }
 
   public Set<DatanodeDetails> getDatanodes() {
-    return datanodes;
+    if (expiryTime > 0) {
```

Comment:
The way this method is now, we will need to iterate the nodes twice - once to remove the expired ones and then once to form the new Set. It would probably be a tiny bit more efficient to just combine the two into one iteration.